File: mobile_app/test/phone_auth_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:mobile_app/core/auth/phone_auth.dart';

// These cases are deliberately the same set as src/lib/phoneAuth.test.js on
// the web. Both apps sign in to the same auth server, so a number normalised
// one way here and another way there becomes two accounts for one person —
// and only one of them has a users row. If either side changes, both must.
void main() {
  const canon = '+919876543210';

  group('toE164', () {
    test('maps every plausible spelling of one number onto one string', () {
      const inputs = [
        '9876543210',
        '09876543210',
        '919876543210',
        '+919876543210',
        '+91 98765 43210',
        '+91-98765-43210',
        '(+91) 98765 43210',
        '  9876543210  ',
        '98765 43210',
        '0091 9876543210',
      ];
      expect(inputs.map(toE164).toSet(), {canon});
    });

    test('refuses what cannot be a mobile rather than guessing', () {
      expect(toE164('98765'), isNull);
      expect(toE164('98765432101234'), isNull);
      expect(toE164(''), isNull);
      expect(toE164('   '), isNull);
      expect(toE164(null), isNull);
      expect(toE164('abcdefghij'), isNull);
      expect(toE164('+'), isNull);
    });

    test('rejects Indian landlines and typos', () {
      expect(toE164('1234567890'), isNull);
      expect(toE164('5876543210'), isNull);
      expect(toE164('6876543210'), '+916876543210');
    });

    test('keeps an explicit foreign country code', () {
      expect(toE164('+14155552671'), '+14155552671');
      expect(toE164('+442071838750'), '+442071838750');
      expect(toE164('+971501234567'), '+971501234567');
    });

    test('does not read a 12-digit number as Indian unless it starts with 91', () {
      expect(toE164('449876543210'), isNull);
    });
  });

  group('formatPhone', () {
    test('formats for reading and leaves the rest untouched', () {
      expect(formatPhone(canon), '+91 98765 43210');
      expect(formatPhone('+14155552671'), '+14155552671');
      expect(formatPhone(''), '');
      expect(formatPhone(null), '');
    });
  });

  group('describeOtpError', () {
    test('names the unknown-number case and keeps the cause', () {
      final original = Exception('Signups not allowed for otp');
      final f = describeOtpError(original);
      expect(f.code, 'NOT_REGISTERED');
      expect(f.message, contains('admin'));
      expect(f.cause, same(original));
    });

    test('separates an expired code from a wrong one', () {
      expect(describeOtpError(Exception('Token has expired')).code, 'EXPIRED');
      expect(describeOtpError(Exception('Invalid token')).code, 'BAD_CODE');
    });

    test('flags rate limiting', () {
      expect(describeOtpError(Exception('too many requests')).code, 'RATE_LIMITED');
    });
  });
}

File: src/lib/phoneAuth.test.js
import { describe, it, expect } from 'vitest';
import { toE164, formatPhone, maskPhone, describeOtpError } from './phoneAuth';

// Supabase matches a stored phone as an exact string. Two spellings of one
// number becoming two accounts is the failure this file guards against, so
// every shape a person might type has to land on the same E.164 value.
describe('toE164', () => {
  const CANON = '+919876543210';

  it('maps every plausible spelling of one number onto one string', () => {
    const inputs = [
      '9876543210',          // bare mobile
      '09876543210',         // STD trunk prefix
      '919876543210',        // country code, no plus
      '+919876543210',       // already E.164
      '+91 98765 43210',     // as printed on a card
      '+91-98765-43210',
      '(+91) 98765 43210',
      '  9876543210  ',      // pasted with whitespace
      '98765 43210',
      '0091 9876543210',     // international 00 prefix
    ];
    const results = new Set(inputs.map((i) => toE164(i)));
    expect([...results]).toEqual([CANON]);
  });

  it('accepts a number given as a number, not a string', () => {
    expect(toE164(9876543210)).toBe(CANON);
  });

  it('refuses what cannot be a mobile rather than guessing', () => {
    // A guess here sends someone's login code to a stranger.
    expect(toE164('98765')).toBeNull();        // too short
    expect(toE164('98765432101234')).toBeNull(); // too long for a local number
    expect(toE164('')).toBeNull();
    expect(toE164('   ')).toBeNull();
    expect(toE164(null)).toBeNull();
    expect(toE164(undefined)).toBeNull();
    expect(toE164('abcdefghij')).toBeNull();
    expect(toE164('+')).toBeNull();
  });

  it('rejects Indian landlines and typos, which cannot receive a WhatsApp OTP', () => {
    // Indian mobiles start 6-9. A 10-digit number starting 0-5 is not one.
    expect(toE164('1234567890')).toBeNull();
    expect(toE164('5876543210')).toBeNull();
    expect(toE164('9876543210')).toBe(CANON); // 9 still fine
    expect(toE164('6876543210')).toBe('+916876543210');
  });

  it('keeps an explicit foreign country code instead of forcing +91', () => {
    expect(toE164('+14155552671')).toBe('+14155552671');
    expect(toE164('+442071838750')).toBe('+442071838750');
    expect(toE164('+971501234567')).toBe('+971501234567');
  });

  it('does not read a 12-digit number as Indian unless it starts with 91', () => {
    expect(toE164('449876543210')).toBeNull();
  });
});

describe('display helpers', () => {
  it('formats for reading but never for lookup', () => {
    expect(formatPhone('+919876543210')).toBe('+91 98765 43210');
    // Anything it cannot format comes back untouched, not mangled.
    expect(formatPhone('+14155552671')).toBe('+14155552671');
    expect(formatPhone('')).toBe('');
    expect(formatPhone(null)).toBe('');
  });

  it('masks enough to confirm the number without exposing it', () => {
    const masked = maskPhone('+919876543210');
    expect(masked).toContain('•');
    expect(masked).not.toBe('+919876543210');
    expect(masked.endsWith('10')).toBe(true);
  });
});

describe('describeOtpError', () => {
  it('names the unknown-number case, which is the one users will hit', () => {
    const e = describeOtpError({ message: 'Signups not allowed for otp', status: 422 });
    expect(e.code).toBe('NOT_REGISTERED');
    expect(e.message).toMatch(/admin/i);
  });

  it('separates an expired code from a wrong one', () => {
    expect(describeOtpError({ message: 'Token has expired' }).code).toBe('EXPIRED');
    expect(describeOtpError({ message: 'Invalid token', status: 401 }).code).toBe('BAD_CODE');
  });

  it('flags rate limiting so the screen does not tell people to retry', () => {
    expect(describeOtpError({ status: 429, message: 'too many requests' }).code).toBe('RATE_LIMITED');
  });

  it('always keeps the original error attached', () => {
    // A message the user can read must not cost the reason a developer needs.
    const original = { message: 'boom', status: 500 };
    const described = describeOtpError(original);
    expect(described.code).toBe('UNKNOWN');
    expect(described.cause).toBe(original);
  });
});
